File: Medium/python/238. Product of Array Except Self.py
'''
Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.
'''


class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        ans = [1]

        prefix = 1
        for i in range(1,len(nums)):
            prefix *= nums[i-1]
            ans.append(prefix)

        suffix = 1
        for i in range(len(nums) -2 ,-1,-1):
            suffix *= nums[i + 1]
            ans[i] = ans[i] * suffix

        return ans

print(Solution().productExceptSelf([1,2,3,4]))

File: Medium/Javascript/238. Product of Array Except Self.js
/**Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation. */

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  const ans = new Array(nums.length).fill(1);

  let prefix = 1;
  for (let i = 1; i < nums.length; i++) {
    prefix *= nums[i - 1];
    ans[i] = prefix;
  }

  let suffix = 1;
  for (let i = nums.length - 2; i >= 0; i--) {
    suffix *= nums[i + 1];
    ans[i] = ans[i] * suffix;
  }

  return ans;
};

console.log(productExceptSelf([1, 2, 3, 4]));
